api/v1/highlights/PKPHighlightsHandler.php

```diff
@@ -0,0 +1,301 @@
+        } else {
+            $collector->withSiteHighlights(Collector::SITE_ONLY);
+        }
+
+        Hook::call('API::highlights::params', [$collector, $slimRequest]);
+
+        $highlights = $collector->getMany();
+
+        return $response->withJson([
+            'itemsMax' => $collector->limit(null)->offset(null)->getCount(),
+            'items' => Repo::highlight()->getSchemaMap()->summarizeMany($highlights)->values(),
+        ], 200);
+    }
+
+    /**
+     * Add a highlight
+     */
+    public function add(SlimRequest $slimRequest, APIResponse $response, array $args): APIResponse
+    {
+        $context = $this->getRequest()->getContext();
+
+        $params = $this->convertStringsToSchema(PKPSchemaService::SCHEMA_HIGHLIGHT, $slimRequest->getParsedBody());
+        $params['contextId'] = $context?->getId();
+        if (!$params['sequence']) {
+            $params['sequence'] = Repo::highlight()->getNextSequence($context?->getId());
+        }
+
+        $errors = Repo::highlight()->validate(null, $params, $context);
+
+        if (!empty($errors)) {
+            return $response->withStatus(400)->withJson($errors);
+        }
+
+        $highlight = Repo::highlight()->newDataObject($params);
+
+        try {
+            $highlightId = Repo::highlight()->add($highlight);
+        } catch (Exception $e) {
+            $highlight = Repo::highlight()->get($highlightId, $context?->getId());
+            Repo::highlight()->delete($highlight);
+            return $response->withStatus(400)->withJson([
+                'image' => __('api.highlights.400.errorUploadingImage')
+            ]);
+        }
+
+        $highlight = Repo::highlight()->get($highlightId, $context?->getId());
+
+        return $response->withJson(Repo::highlight()->getSchemaMap()->map($highlight), 200);
+    }
+
+    /**
+     * Edit a highlight
+     */
+    public function edit(SlimRequest $slimRequest, APIResponse $response, array $args): APIResponse
+    {
+        $context = $this->getRequest()->getContext();
+
+        $highlight = Repo::highlight()->get((int) $args['highlightId'], $context?->getId());
+
+        if (!$highlight) {
+            return $response->withStatus(404)->withJsonError('api.highlights.404.highlightNotFound');
+        }
+
+        $params = $this->convertStringsToSchema(PKPSchemaService::SCHEMA_HIGHLIGHT, $slimRequest->getParsedBody());
+        $params['id'] = $highlight->getId();
+
+        // Not allowed to change the context of a highlight through the API
+        unset($params['contextId']);
+
+        $errors = Repo::highlight()->validate($highlight, $params, $context);
+
+        if (!empty($errors)) {
+            return $response->withStatus(400)->withJson($errors);
+        }
+
+        try {
+            Repo::highlight()->edit($highlight, $params);
+        } catch (Exception $e) {
+            Repo::highlight()->delete($highlight);
+            return $response->withStatus(400)->withJson([
```

Comment:
Is this the only possible exception? I would worry that other errors will be presented as image upload errors, which might be hard to debug.

Reply:
Good point, I've introduced a `PKP\core\exceptions\StoreTemporaryFileException` in both this PR and the announcements one (which does the same).